torchqdynamics/tensor_types.py

```diff
@@ -22,30 +27,99 @@
 TDOperatorLike = Union[OperatorLike, Callable[[float], OperatorLike]]
 
 
-def to_tensor(x: OperatorLike | list[OperatorLike] | None) -> Tensor:
+def to_tensor(
+    x: OperatorLike | list[OperatorLike] | None,
+    *,
+    dtype: torch.dtype | None = None,
+    device: torch.device | None = None,
+    is_complex: bool = False,
+) -> Tensor:
     """Convert a `OperatorLike` object or a list of `OperatorLike` object to a PyTorch
     tensor.
 
     Args:
         x: QuTiP quantum object or NumPy array or Python list or PyTorch tensor or list
            of these types. If `None` or empty list, returns an empty tensor of size (0).
+        dtype: Data type of the returned tensor.
+        device: Device on which the returned tensor is stored.
 
     Returns:
         PyTorch tensor.
     """
+    if is_complex:
+        dtype = cdtype(dtype)
+
     if x is None:
-        return torch.tensor([])
+        return torch.tensor([], dtype=dtype, device=device)
     if isinstance(x, list):
         if len(x) == 0:
-            return torch.tensor([])
-        return torch.stack([to_tensor(y) for y in x])
+            return torch.tensor([], dtype=dtype, device=device)
+        return torch.stack([to_tensor(y, dtype=dtype, device=device) for y in x])
     if isinstance(x, Qobj):
-        return from_qutip(x)
+        return from_qutip(x, dtype=dtype, device=device)
     elif isinstance(x, get_args(TensorLike)):
-        return torch.as_tensor(x)
+        return torch.as_tensor(x, dtype=dtype, device=device)
     else:
         raise TypeError(
             f'Input of type {type(x)} is not supported. `to_tensor` only '
             'supports QuTiP quantum object, NumPy array, Python list or PyTorch tensor '
             'or list of these types.'
         )
+
+
+def cdtype(
+    dtype: torch.complex64 | torch.complex128 | None = None,
+) -> torch.complex64 | torch.complex128:
+    if dtype is None:
+        # the default dtype for complex tensors is determined by the default
+        # floating point dtype (torch.complex128 if default is torch.float64,
+        # otherwise torch.complex64)
+        if torch.get_default_dtype() is torch.float64:
+            return torch.complex128
+        else:
+            return torch.complex64
+    elif dtype not in (torch.complex64, torch.complex128):
```

Comment:
If you use `to_tensor(x, dtype=torch.float32, is_complex=True)` this will raise an error. Is this the behavior we want? Or do we want to convert float32 to complex64 in this case?

Reply:
Yes that's on purpose, I think one should always specify a complex type for a complex tensor, and we should avoid automatic conversions. But it can be discussed, typically `mesolve` will fail if someone passes `dtype=torch.float32`, he would have to pass `dtype=torch.complex64`, but I think it's better and more clear like that.